mesh/submesh/submesh.hpp

```diff
@@ -0,0 +1,201 @@
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_SUBMESH
+#define MFEM_SUBMESH
+
+#include "../mesh.hpp"
+#include "transfermap.hpp"
+#include <unordered_map>
+
+namespace mfem
+{
+/**
+ * @brief Subdomain representation of a topological parent in another Mesh.
+ *
+ * SubMesh is a subdomain representation of a Mesh defined on it's parents
+ * attributes. The current implementation creates either a domain or surface
+ * subset of the parents Mesh and reuses the parallel distribution.
+ *
+ * The attributes are taken from the parent. That means if a volume is extracted
+ * from a volume, it has the same domain attribute as the parent. It's boundary
+ * attributes are generated (there will be one boundary attribute 1 for all of
+ * the boundaries).
+ *
+ * If a surface is extracted from a volume, the boundary attribute from the
+ * parent is assigned to be the new domain attribute. It's boundary attributes
+ * are generated (there will be one boundary attribute 1 for all of the
+ * boundaries).
+ *
+ * For more customized boundary attributes, the resulting SubMesh has to be
+ * postprocessed.
+ */
+class SubMesh : public Mesh
+{
+public:
+   /// Indicator from which part of the parent Mesh the SubMesh is created.
+   enum From
+   {
+      Domain,
+      Boundary
+   };
+
+   static const int GENERATED_ATTRIBUTE = 900;
+
+   SubMesh() = delete;
+
+   /**
+    * @brief Create a domain SubMesh from it's parent.
+    *
+    * The SubMesh object expects the parent Mesh object to be valid for the
+    * entire object lifetime. The @a domain_attributes have to mark exactly one
+    * connected subset of the parent Mesh.
+    *
+    * @param[in] parent Parent Mesh
+    * @param[in] domain_attributes Domain attributes to extract
+    */
+   static SubMesh CreateFromDomain(const Mesh &parent,
+                                   Array<int> domain_attributes);
+
+   /**
+   * @brief Create a surface SubMesh from it's parent.
+   *
+   * The SubMesh object expects the parent Mesh object to be valid for the
+   * entire object lifetime. The @a boundary_attributes have to mark exactly one
+   * connected subset of the parent Mesh.
+   *
+   * @param[in] parent Parent Mesh
+   * @param[in] boundary_attributes Boundary attributes to extract
+
+   */
+   static SubMesh CreateFromBoundary(const Mesh &parent,
+                                     Array<int> boundary_attributes);
+
+   /**
+    * @brief Get the parent Mesh object
+    *
+    */
+   const Mesh* GetParent() const
```

Comment:
Does this have to return a pointer? I think a reference with value semantics is cleaner.

Reply:
I agree but it makes some functions (like recursive lookups) very clean to write.